src/ray/gcs/node_state_accessor.h

```diff
@@ -0,0 +1,124 @@
+#ifndef RAY_GCS_NODE_STATE_ACCESSOR_H
+#define RAY_GCS_NODE_STATE_ACCESSOR_H
+
+#include <boost/optional.hpp>
+#include <map>
+#include <memory>
+#include <mutex>
+#include <vector>
+#include "ray/common/id.h"
+#include "ray/gcs/callback.h"
+#include "ray/gcs/entry_change_notification.h"
+#include "ray/gcs/tables.h"
+
+namespace ray {
+
+namespace gcs {
+
+class RedisGcsClient;
+class NodeStateCache;
+
+/// \class NodeStateAccessor
+/// NodeStateAccessor class encapsulates the implementation details of
+/// reading, writing and subscribing of node's information (immutable fields like id,
+/// and mutable fields like runtime state).
+class NodeStateAccessor {
+ public:
+  explicit NodeStateAccessor(RedisGcsClient *client_impl);
+
+  ~NodeStateAccessor() {}
+
+  /// Register local node to GCS synchronously, and begins subscription to all nodes
+  /// from GCS (the node informations will be cached).
+  ///
+  /// \param node_info The information of node to register to GCS.
+  /// \return Status
+  Status RegisterSelf(const GcsNodeInfo &local_node_info);
+
+  /// Cancel registration of local node to GCS synchronously, and cancel subscription
+  /// to all nodes from GCS.
+  ///
+  /// \return Status
+  Status UnregisterSelf();
+
+  /// Get id of local node which registered by 'RegisterSelf'.
+  ///
+  /// \return ClientID
+  const ClientID &GetSelfId() const;
+
+  /// Get information of local node which registered by 'RegisterSelf'.
+  ///
+  /// \return GcsNodeInfo
+  const GcsNodeInfo &GetSelfInfo() const;
+
+  /// Cancel registration of a node to GCS asynchronously.
+  ///
+  /// \param node_id The ID of node that to be unregistered.
+  /// \param callback Callback that will be called when unregistration is complete.
+  /// \return Status
+  Status AsyncUnregister(const ClientID &node_id, const StatusCallback &callback);
+
+  /// Get information of all nodes from GCS asynchronously.
+  ///
+  /// \param callback Callback that will be called after lookup finishes.
+  /// \return Status
+  Status AsyncGetAll(const MultiItemCallback<GcsNodeInfo> &callback);
+
+  /// Subscribe to node add or node remove from GCS.
+  ///
+  /// \param subscribe Callback that will be called if a node is
+  /// added or a node is removed.
+  /// \param done Callback that will be called when subscription is complete.
+  /// \return Status
+  Status AsyncSubscribeToNodeChange(
+      const SubscribeCallback<ClientID, GcsNodeInfo> &subscribe,
+      const StatusCallback &done);
+
+  /// Get nodes cache. The cache stores node information subscribed from GCS.
+  /// It provides read-only interfaces such as `Get`.
+  ///
+  /// \return NodeStateCache
+  NodeStateCache &Cache() const;
```

Comment:
I prefer not to expose 'cache' in the API. Can we just move the methods from `NodeStateCache` to this class?
As we've subscribed to the changes, the local cache should be mostly up-to-date. So we can just consider `Get` as a sync function.

Reply:
First of all, the `cache` is refreshed only after calling `SubscribeToNodeChange`.  Those process that don't subscribe nodes information from GCS,  the cache won't be refreshed, and the only way to get node information is read from GCS. If we rename this method to `Get`, it would be so confused and might be incorrect used by user.

Secondly, if the logic is  find in cache first and if not found get from GCS, then we can use `Get` instead . `Cache.Get()` only means get from `cache`, that's all we need now.

Reply:
>  find in cache first and if not found get from GCS, then we can use Get instead

This would be better,  cache should only be an internal optimization and don't need to be exposed in API.